### PR TITLE
feat(server): Add user.id and user.segment to dynamic rules

### PR DESCRIPTION
**Pull request by RaduW**
This PR adds support for `user.id` and `user.segement` in sampling rules.
Both `TraceContext.user` and `Event.user` now have `id` and `segment` 
optional fields.

#skip-changelog

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+**Features**:
+
+- Add `user.id` and `user.segment` to sampling rules. ([#959](https://github.com/getsentry/relay/pull/959))
```
Comment (jan-auer) [inline CHANGELOG.md]: With all changes related to dynamic sampling, let's please keep them in "Internal", as this is still under development and not an available feature yet.

File: relay-sampling/src/lib.rs [hunk 2]
```diff
@@ -472,6 +492,15 @@ impl SamplingConfig {
     }
 }
 
+/// The User related information in the trace context
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct TraceUserContext {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub segment: Option<String>,
```
Comment (jan-auer) [inline relay-sampling/src/lib.rs]: What if you make this non-optional? You still check for empty strings, but can avoid one more indirection. We should also test for the case where an empty string is sent through `id` or `segment` and ensure it is ignored.

```suggestion
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub segment: String,
```

Comment (RaduW) [reply relay-sampling/src/lib.rs]: ok, made segment and id non optional (and check for empty string while matching)

File: CHANGELOG.md (hunk 1 above)
Comment (RaduW) [reply CHANGELOG.md]: ok

**Review by jan-auer — APPROVED** [review]
